### ISSUE TITLE
Array within a dict throws a "list assignment index out of range" exception

### ISSUE THREAD

**Issue by kevinhaeni** [stefankoegl/python-json-patch#74]
Hello, we have encountered the following issue when adding a value to an array within a dictionary:

```
        old = {'key': [{'someNumber': 0, 'someArray': [1, 2, 3]}]}
        new = {'key': [{'someNumber': 0, 'someArray': [1, 2, 3, 4]}]}
        patches = jsonpatch.make_patch(old,new)
```

Make_patch raises a "list assignment index out of range" exception. Any idea?

I have quickly checked with a debugger and it seems that the make_patch originally generates a correct 'add' operation, however, the optimzer creates the following operation:

`[{"value": 4, "path": "/key/0/someArray/3", "op": "replace"}]
`
which is of course not correct and might result in the above error when you apply the operations to the old object:
`    new = patch.apply(src)`

any idea how to fix this? As a workaround, I have just disabled optimizations in the first from_diff call. Is that a possible workaround?

**Comment by cumgun** [comment]
+1

I can reproduce the issue. Not sure, but this issues seems to be related to #36 somehow. Here's a minimal example which reproduces the issue:

```python
import jsonpatch
old = {'a': [{'b': []}]}
new = {'a': [{'b': [1]}]}
patches = jsonpatch.make_patch(old, new)
patches.apply(old)
```

In this example, `jsonpatch.make_patch` creates the JSON patch `[{'op': 'replace', 'path': '/a/0/b/0', 'value': 1}]`, which is obviously erroneous.

**Comment by stefankoegl** [comment]
I have just released jsonpatch 1.20 which fixes the issue you describe. In addition I have added a test for this case in 04596b810baf178ee3799c72c03bfe1990400418.

Thanks for reporting!
